armi/plugins.py

```diff
@@ -600,10 +600,77 @@ def defineSystemBuilders() -> Dict[str, Callable[[str], "Composite"]]:
         """
 
 
-def getNewPluginManager() -> pluginManager.ArmiPluginManager:
+class UserPlugin(ArmiPlugin):
     """
-    Return a new plugin manager with all of the hookspecs pre-registered.
+    This is a variation on the usual ArmiPlugin that is meant to be read in during run time,
+    from a hard-coded path in a settings file, or the like.
+
+    This is obviously a more limited use-case than the usual ArmiPlugin, as those are meant
+    to be defined at import time, instead of run time. As such, this class has some built-in
+    tooling to limit how these run-time plugins are used. They are meant to be more limited.
+
+    NOTE: The usual ArmiPlugin is much more flexible, if the UserPlugin does not support what
+    you want to do, just use an ArmiPlugin.
     """
+
+    def __init__(self, *args, **kwargs):
+        ArmiPlugin.__init__(self, *args, **kwargs)
+        self.__enforceLimitations()
+
+    def __enforceLimitations(self):
+        """
+        This method is designed to enforce that UserPlugins are more limited
+        than regular ArmiPlugins.
+        UserPlugins are different from regular plugins in that they can be defined during
+        a run, and as such, we want to limit how flexible they are, so we can correctly
+        corral their side effects during a run.
+        """
+        if issubclass(self.__class__, UserPlugin):
+            assert (
+                len(self.__class__.defineParameterRenames()) == 0
+            ), "UserPlugins cannot define parameter renames"
+            assert (
+                len(self.__class__.defineSettings()) == 0
+            ), "UserPlugins cannot define new Settings"
```

Comment:
consider adding an actionable suggestion in these message, like "...consider using an ArmiPlugin instead"

Reply:
Done